// @atlaskit/button/dist/es2019/new-button/variants/default/button.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import UNSAFE_PRESSABLE from '@atlaskit/primitives/pressable';
import useDefaultButton from './use-default-button';
/**
 * __Button__
 *
 * A button triggers an event or action.
 *
 * - [Examples](https://atlassian.design/components/button/examples)
 * - [Code](https://atlassian.design/components/button/code)
 * - [Usage](https://atlassian.design/components/button/usage)
 */
const Button = /*#__PURE__*/React.memo( /*#__PURE__*/React.forwardRef(function Button({
  analyticsContext,
  autoFocus,
  appearance,
  spacing,
  isDisabled,
  isSelected,
  iconBefore,
  UNSAFE_iconAfter_size,
  iconAfter,
  UNSAFE_iconBefore_size,
  children,
  shouldFitContainer,
  interactionName,
  overlay,
  onClick,
  onMouseDownCapture,
  onMouseUpCapture,
  onKeyDownCapture,
  onKeyUpCapture,
  onTouchStartCapture,
  onTouchEndCapture,
  onPointerDownCapture,
  onPointerUpCapture,
  onClickCapture,
  type = 'button',
  testId,
  ...rest
}, ref) {
  const baseProps = useDefaultButton({
    analyticsContext,
    appearance,
    autoFocus,
    buttonType: 'button',
    children,
    iconBefore,
    UNSAFE_iconBefore_size,
    iconAfter,
    UNSAFE_iconAfter_size,
    interactionName,
    isDisabled,
    isSelected,
    onClick,
    onMouseDownCapture,
    onMouseUpCapture,
    onKeyDownCapture,
    onKeyUpCapture,
    onTouchStartCapture,
    onTouchEndCapture,
    onPointerDownCapture,
    onPointerUpCapture,
    onClickCapture,
    overlay,
    ref,
    shouldFitContainer,
    spacing
  });
  return /*#__PURE__*/React.createElement(UNSAFE_PRESSABLE
  // TODO: Remove spread props
  , _extends({}, rest, {
    ref: baseProps.ref,
    xcss: baseProps.xcss,
    isDisabled: baseProps.isDisabled,
    onClick: baseProps.onClick,
    onMouseDownCapture: baseProps.onMouseDownCapture,
    onMouseUpCapture: baseProps.onMouseUpCapture,
    onKeyDownCapture: baseProps.onKeyDownCapture,
    onKeyUpCapture: baseProps.onKeyUpCapture,
    onTouchStartCapture: baseProps.onTouchStartCapture,
    onTouchEndCapture: baseProps.onTouchEndCapture,
    onPointerDownCapture: baseProps.onPointerDownCapture,
    onPointerUpCapture: baseProps.onPointerUpCapture,
    onClickCapture: baseProps.onClickCapture,
    type: type,
    testId: testId
  }), baseProps.children);
}));
Button.displayName = 'Button';
export default Button;
